packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -1764,28 +1764,14 @@ class OpacityLayer extends ContainerLayer {
   set alpha(int? value) {
     assert(value != null);
     if (value != _alpha) {
+      if (value == 255 || _alpha == 255) {
+        engineLayer = null;
```

Comment:
The type of the engineLayer is specific to the implementation of the addToScene method. Rather than null it here where the reason for the nullification isn't immediately obvious, deal with the wrong type down in addToScene.

Basically in addToScene, only pass along the oldLayer if it is of the right type and it will get replaced naturally.

Reply:
Because addToScene is on a hot path, and will require type checks every time that hot path gets hit, whereas we know what the type will be based ont he value of the opacity.

Reply:
A single `is` is going to completely derail our performance?
Compared to the improved maintainability by not having multiple functions have to deal with an implementation detail of one of them?

addToScene has to do an `as` anyway which implicitly includes an `is`, but throws if it fails rather than just substituting null which is the logical reaction to the type being wrong. Note that with the `is` you likely don't even need the `as` because the runtime is smart enough to elide them.
You are already paying the price of type checking the parameter in addToScene, why complicate it?

Reply:
So I figured I don't have a benchmark to back this up and I can just make this change. However, there's another complication here. What I really want is it to be an error to call `addToScene` with an incorrect engineLayer, because you cannot mutate the `engineLayer` during addToScene, and if we got called `addToScene` without nulling out the no-longer-needed engineLayer, we've made a mistake that wastes memory.

I'm happy to update the documentation on this if it's not already updated, but I think the asserts and mutations are correct this way, and we'd risk recurisvely calling `addToScene` if we start trying to mutate the engineLayer there.

Reply:
How is "this object wants to decide on the fly which type of engine layer to use when it adds to the scene" compatible with "addToScene is not allowed to mutate the engineLayer and we will make that an error"?

Philosophically those two concepts are at odds with each other. You can't justify one without pointing out that the other is or has an invalid assumption.

Adding code to prevent a conflict between them is a workaround and points out that the design decisions had some cases they didn't really address.

You are basically nulling out the field to destroy the evidence that the engine layer is actually dynamic here.

And if there is a reason for it to be dynamic here, then what is the value of a globally enforced assertion that it isn't.

Reply:
(None of this is blocking, btw, I'm just a little skeptical over the design constraints that led to this implementation...)